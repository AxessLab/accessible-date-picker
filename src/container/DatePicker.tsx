import React, { useCallback, useEffect, useState } from "react";
import moment from "moment";
import styles from "../styles/stylesDatePicker.css";

import CalendarIcon from "../components/CalendarIcon";
import MonthPicker from "../components/MonthPicker";
import DaysHeading from "../components/DaysHeading";
import DatesOfMonth from "../components/DatesOfMonth";
import keyDownHandler from "../utility/keyDownHandler";
import onChangeHandler from "../utility/onChangeHandler";

interface IDatePickerProps {
  applicationMode?: boolean;
  value: string;
  setValue: React.Dispatch<React.SetStateAction<string>>;
  dateFormat: string;
  validation: boolean;
}

interface IClickedDate {
  date?: number;
  month?: number;
  year?: number;
}

interface IDateObject {
  year: number;
  month: number;
  dates: number;
}

interface IIsClicked {
  buttonId: string,
  selected: boolean
}

const DatePicker: React.FC<IDatePickerProps> = (props) => {
  const { value, setValue, dateFormat, validation } = props
  const applicationMode = props.applicationMode ? true : false;
  const [showCalendar, setShowCalendar] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");
  const [clickedDate, setClickedDate] = useState<IClickedDate>({});
  const [dateObject, setDateObject] = useState<IDateObject>({
    year: +moment().year(),
    month: +moment().month(),
    dates: +moment().daysInMonth(),
  });
  const [isClicked, setIsClicked] = useState<IIsClicked>({
    buttonId: "",
    selected: false
  });

  useEffect(() => {
    setClickedDate({
      ...clickedDate,
      year: +dateObject.year,
      month: +dateObject.month + 1
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dateObject]);

  useEffect(() => {
    const dateInputValueHandler = () => {
      if (clickedDate.date) {
        const selectedFormDateValue = moment(`${clickedDate.year}-${clickedDate.month}-${clickedDate.date}`, "YYYY-MM-DD").format(dateFormat);

        setValue(selectedFormDateValue);
      }
    };
    dateInputValueHandler();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [clickedDate]);

  const showCalendarHandler = () => {
    setShowCalendar(!showCalendar);
  };

  const escCalendar = (event: React.KeyboardEvent) => {
    if (event.key === "Escape") {
      setShowCalendar(false);
    }
  };

  const onKeyDown = useCallback((event) => { keyDownHandler(event, dateObject.dates, setShowCalendar) }, [dateObject.dates]);
  const applicationKeyHandler = (applicationMode: boolean) => {
    if (applicationMode) {
      window.addEventListener('keydown', onKeyDown);
    }
    else {
      window.removeEventListener('keydown', onKeyDown)
    }
  };

  return (
    <>
      <div onKeyDown={(e) => escCalendar(e)}>
        <label htmlFor="date-picker-input" className={styles.label} aria-label="enter date in the following format">{dateFormat}</label><br />
        <button className={styles.iconButton} aria-label={showCalendar ? "select to close calendar" : "select to open calendar"} type="button" onClick={showCalendarHandler}><CalendarIcon /></button>
        <input className={styles.inputField} id="date-picker-input" type="text" aria-label={value.length > 1 ? "entered date value is" : `enter date in format ${dateFormat}`} autoComplete="off" value={value}
          onChange={(e) => onChangeHandler(e, dateFormat, validation, setValue, setErrorMessage, setClickedDate, setIsClicked)} />
        {errorMessage && <div aria-live="assertive" role="alert"><p style={{ color: "#871111", padding: "4px" }} >{errorMessage}</p></div>}
      </div>
      <div className={showCalendar ? styles.calendarContainer : styles.hidden} {...(applicationMode ? { role: "application" } : {})} >
        <MonthPicker
          currentDate={dateObject}
          clickedDate={clickedDate}
          setDateObject={setDateObject} />
        <table id='calendar-table' className={styles.calendarTableContainer} role="presentation" onKeyDown={() => applicationKeyHandler(applicationMode)} >
          <DaysHeading />
          <DatesOfMonth
            applicationMode={applicationMode}
            year={dateObject.year}
            month={dateObject.month}
            datesOfMonth={dateObject.dates}
            setClickedDate={setClickedDate}
            showCalendarHandler={showCalendarHandler}
            isClicked={isClicked}
            setIsClicked={setIsClicked}
            setErrorMesage={setErrorMessage}
          />
        </table>
      </div>
    </>
  );
};

export default DatePicker;