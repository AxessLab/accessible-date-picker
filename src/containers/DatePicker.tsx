import React, { useCallback, useEffect, useState } from "react";
import moment from "moment";
import styles from "../styles/stylesDatePicker.css";

import CalendarIcon from "../components/CalendarIcon";
import MonthPicker from "../components/MonthPicker";
import DaysHeading from "../components/DaysHeading";
import DatesOfMonth from "../components/DatesOfMonth";
import { keyDownHandler } from "../utility/functions";

interface CalendarProps {
  applicationMode?: boolean;
  datePickerFormValue: string;
  setDatePickerFormValue: (value: string) => void;
}

interface ClickedDate {
  year?: number | null;
  month?: number | null;
  date?: number | null
}

const Calendar: React.FC<CalendarProps> = (props) => {
  const { datePickerFormValue, setDatePickerFormValue } = props
  const applicationMode = props.applicationMode ? true : false;
  const [showCalendar, setShowCalendar] = useState(false);
  const [clickedDate, setClickedDate] = useState<ClickedDate>({});
  const [dateObject, setDateObject] = useState({
    year: +moment().year(),
    month: +moment().month(),
    dates: +moment().daysInMonth(),
  });

  useEffect(() => {
    setClickedDate({
      ...clickedDate,
      year: dateObject.year,
      month: dateObject.month + 1
    });

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dateObject]);

  useEffect(() => {
    const dateInputValueHandler = () => {
      if (clickedDate.date) {
        const selectedFormDateValue = moment(`${clickedDate.year}-${clickedDate.month}-${clickedDate.date}`, "YYYY-MM-DD").format('MM/DD/YYYY');

        setDatePickerFormValue(selectedFormDateValue);
      }
    };

    dateInputValueHandler();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [clickedDate]);

  const showCalendarHandler = () => {
    setShowCalendar(!showCalendar);
  };

  const onChangeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    setDatePickerFormValue(event.target.value);
  };

  const onKeyDown = useCallback((event) => { keyDownHandler(event, dateObject.dates) }, [dateObject.dates]);
  const applicationKeyHandler = (applicationMode: boolean) => {
    if (applicationMode) {
      window.addEventListener('keydown', onKeyDown);
    }
    else {
      window.removeEventListener('keydown', onKeyDown)
    }
  };

  return (
    <>
      <div>
        <button className={styles.iconButton} aria-label="click for calendar" type="button" onClick={showCalendarHandler}><CalendarIcon /></button>
        <input className={styles.inputField} id="date-picker-input" type="text" aria-label="date input" placeholder='MM/DD/YYYY' autoComplete="off" value={datePickerFormValue} onChange={(e) => onChangeHandler(e)} />
      </div>
      <div className={showCalendar ? styles.calendarContainer : styles.hidden} {...(applicationMode ? { role: "application" } : {})}>
        <MonthPicker
          month={dateObject.month}
          year={dateObject.year}
          setDateObject={setDateObject} />
        <table id='calendar-table' className={styles.calendarTableContainer} role="presentation" onKeyDown={() => applicationKeyHandler(applicationMode)} >
          <DaysHeading />
          <DatesOfMonth
            applicationMode={applicationMode}
            year={dateObject.year}
            month={dateObject.month}
            datesOfMonth={dateObject.dates}
            setClickedDate={setClickedDate}
            showCalendarHandler={showCalendarHandler}
          />
        </table>
      </div>
    </>
  );

};

export default Calendar;