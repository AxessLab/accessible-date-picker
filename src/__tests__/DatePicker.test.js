import React from "react";
import { shallow, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import DatePicker from "../container/DatePicker";

configure({ adapter: new Adapter() });

const mockValue = ""
const mockSetValue = jest.fn();
const wrapper = shallow(<DatePicker value={mockValue} setValue={mockSetValue} dateFormat="YYYY/MM/DD" />);

test("datePicker changes input state value", () => {
    wrapper.find('input').simulate('change', { target: { name: 'date-picker-input', value: "2020/01/02" } });

    expect(mockSetValue).toHaveBeenCalledWith("2020/01/02");

    expect(toJson(wrapper)).toMatchSnapshot();
});
