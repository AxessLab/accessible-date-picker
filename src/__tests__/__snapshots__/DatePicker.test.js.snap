// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datePicker changes input state value 1`] = `
<ThemeProvider
  theme={
    Object {
      "palette": Object {
        "primary": "#f5f5f5",
        "secondary": "#2b4450",
        "tertiary": "#871111",
      },
      "spacing": Array [
        "0px",
        "4px",
        "8px",
        "16px",
        "32px",
        "64px",
      ],
    }
  }
>
  <div
    onKeyDown={[Function]}
  >
    <label
      aria-label="enter date in the following format"
      className="labelDatePicker-0-2-1"
      htmlFor="date-picker-input"
    >
      YYYY/MM/DD
    </label>
    <br />
    <button
      aria-label="select to open calendar"
      className="iconButton-0-2-3"
      onClick={[Function]}
      type="button"
    >
      <CalendarIcon />
    </button>
    <input
      aria-label="enter date in format YYYY/MM/DD"
      autoComplete="off"
      className="inputFieldDatePicker-0-2-2"
      id="date-picker-input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="hiddenCalendar-0-2-6"
  >
    <MonthPicker
      clickedDate={
        Object {
          "date": 2,
          "month": 1,
          "year": 2020,
        }
      }
      currentDate={
        Object {
          "dates": 28,
          "month": 1,
          "year": 2021,
        }
      }
      setDateObject={[Function]}
    />
    <table
      className="calendarTableContainer-0-2-5"
      id="calendar-table"
      onKeyDown={[Function]}
      role="presentation"
    >
      <DaysHeading />
      <DatesOfMonth
        applicationMode={false}
        datesOfMonth={28}
        isClicked={
          Object {
            "buttonId": "button-2",
            "selected": true,
          }
        }
        month={1}
        setClickedDate={[Function]}
        setErrorMesage={[Function]}
        setIsClicked={[Function]}
        showCalendarHandler={[Function]}
        year={2021}
      />
    </table>
  </div>
</ThemeProvider>
`;
